beyond-all-reason/Beyond-All-Reason
# Unit Market - Widget and Gadget backend.

### Work done
Added a LuaUI widget and LuaRules gadget to facilitate unit trading between players. The LuaUI widget provides a user interface for players to offer units for sale, while the LuaRules gadget manages the exchange process. Players can now initiate trades by offering units for sale, and others can purchase them by alt+double-clicking on the offered units.

#### Addresses Issue(s)
Implemented a feature allowing players to set their units for sale effortlessly, without requiring any additional action from the seller to complete the sale for metal. Additionally, the buyer is relieved from the task of manually sending metal for the purchase.

#### Setup
This is disabled by default. You need to set the modoption unit_market to true. It is in the extra tab. So no extra steps necessary to make it work. See below on how to test it.

#### Test steps
Route A "single player test":
1. `AllyAIsalesEverything` is true by default. Otherwise change it to `true`.
2. Add AI to your team.
3. Have AI build something. (/cheat and /godmode 3 can help you, /godmode 0 when you are finished)
4. Hover mouse over that unit, make sure you have enough metal, hold ALT and double-click on the unit.
5. The unit is now yours.

Route B "multi player match":
1. Have at least two players on the team.
2. Player A offers a unit for sale, they select unit and, for example they press "Sell Unit" button or they have a hotkey (/bind alt+c /sell_unit) or type /sell_unit in chat and hit enter.
3. Player B makes sure they have enough metal to buy unit. Hover mouse over that unit, hold ALT and double-click on the unit.
4. The unit now belongs to Player B and Player A is "metalCost" metal richer.

### Extra info
This is the proper implementation of [this proof-of-concept widget](https://discord.com/channels/549281623154229250/1134829322390212698).

If you are the seller you should probably make sure you have enough metal storage, you can overflow. I can code "escrow" storage though, but not sure if its necessary.

If you are widget maker you can use this gadget capability to facilitate unit trading. For example, you can set it so all your idle t1/t2 construction bots will automatically be set for sale. You can also set that the moment you have your first afus - you set all your solar/winds/etc are for sale too. So any player who is catching up on eco can buy them right way without the need to build them on their own.

Let me know if you want a separate feature so that, if someone reclaims your unit or building - they pay you for the cost of that building. (anti-grief)

The gadget was meant to be as simple as possible, most of the leg work is done by widget (either this or others can make their own versions or add extra logic to auto-set certain units for sale if they are idle for example).

Update: AI trading is enabled by default. Also for better coop experience - AI will keep tabs on how much it owes you. So, for example, if you gift your factory to the AI, you can buy AI's factory with a discount or for free, depending on the metal cost difference between two factories.

### A bit of info for widget coding
You do not need to supply price at all, or you can supply 0. Either way gadget will automatically set the price to metalCost of the unit. If you want to undo-sell status - just send the offer to sell again. Gadget will flip the unit back to 'not selling' for you.
```lua
local function OfferToSell(unitID)
    Spring.SendLuaRulesMsg("unitOfferToSell " .. unitID) -- Tell gadget we are offering unit for sale
end
local function OfferToBuy(unitID)
    Spring.SendLuaRulesMsg("unitTryToBuy " .. unitID) -- Tell gadget we are buying (or trying to)
end
```
In order to listen for incoming trade offers and done deals you'll need:
```lua
function widget:UnitSale(unitID, price, msgFromTeamID) -- unit, price, unitTeam
    -- your code
end
function widget:UnitSold(unitID, price, old_ownerID, msgFromTeamID) -- unit, price, oldTeamID, newTeamID
    -- your code
end
```
In order to check if unit is currently being on sale at any time you can do this:
`local price = Spring.GetUnitRulesParam(unitID, "unitPrice")`
If price is zero - it's not being sold by its owner.
So knowing this - it should be trivial to, for example, code something to immediately buy metalmakers if you have spare metal and spare energy if anyone places them on sale, just an example. And the other way - the moment your afus is up, you can bulk send on sale all your windgens and solars. Just don't try to scam someone by selling them unit and then reclaiming it right away...

### Screenshots:
https://imgur.com/a/EkWYJkn
https://imgur.com/a/DKGS0wE - (old button, deprecated)
https://imgur.com/a/jLsjd9F - new button is here now.
